Handle SIGHUP to cause the log to rotate/get a new one manually

andybarry opened the pull request:
The zcm-logger help says that SIGHUP should cause the log to rotate, but that didn't work for me.

I have implemented that functionality here.

jbendes commented:
This looks great thanks for the contribution! What are your thoughts on behavior if `args.rotate <= 0`?

andybarry commented:
Good point.  Perhaps we only support SIGHUP if rotate or increment is used?  I'll implement that and add a print if SIGHUP is received and we're not in that configuration.

andybarry commented:
I've updated the PR with a check that ignores SIGHUP if the user hasn't specified `--increment` or `--rotate`.  I also was a bit confused about how to not have `--increment` specified, so I changed the error message there to add some more context, hopefully that makes sense to you.

jbendes commented on tools/cpp/logger/main.cpp:
```diff
@@ -207,7 +208,7 @@ struct Args
 
         if (rotate > 0 && auto_increment) {
-            cerr << "ERROR.  --increment and --rotate can't both be used" << endl;
+            cerr << "ERROR.  --increment and --rotate can't both be used. Note that if you don't want --increment, you must specify a log filename." << endl;
```
Can we make sure to adhere to max line length here? Run `zcm-logger -h` to see what I mean

jbendes commented on tools/cpp/logger/main.cpp:
```diff
@@ -580,6 +581,19 @@ struct Logger
         }
 
+        // Did we get a SIGHUP and the user wants a new logfile?
+        if (got_sighup && (args.auto_increment || (args.rotate > 0))) {
```
Can just check `got_sighup`

jbendes commented on tools/cpp/logger/main.cpp:
```diff
@@ -683,6 +703,7 @@ int main(int argc, char *argv[])
     signal(SIGINT,  sighandler);
     signal(SIGQUIT, sighandler);
     signal(SIGTERM, sighandler);
+    signal(SIGHUP,  sighup_handler);
```
You can register this handler only in the cases you want to actually handle the signal. Specifically `args.auto_increment || args.rotate > 0`

andybarry commented:
All updated, note there is a slight behavior change in that if we don't register SIGHUP handler, the logger will quit when it receives a SIGHUP.

jbendes commented:
Ahh good point. Let's preserve the old behavior with the registering the signal. But you can just not set the variable in the callback unless it's relevant to
